maven-core/src/main/java/org/apache/maven/plugin/internal/DefaultMavenPluginManager.java

```diff
@@ -492,6 +492,21 @@ private void discoverPluginComponents( final ClassRealm pluginRealm, Plugin plug
 
         ClassRealm pluginRealm = pluginDescriptor.getClassRealm();
 
+        if ( pluginRealm == null )
+        {
+            try
+            {
+                setupPluginRealm( pluginDescriptor, session, null, null, null );
+            }
+            catch ( PluginResolutionException e )
+            {
+                String msg = "Cannot setup plugin realm [mojoDescriptor=" + mojoDescriptor.getId()
+                        + ", pluginDescriptor=" + pluginDescriptor.getId() + "]";
+                throw new PluginConfigurationException( pluginDescriptor, msg, e );
+            }
+            pluginRealm = pluginDescriptor.getClassRealm();
+        }
+
         if ( logger.isDebugEnabled() )
```

Comment:
@maximilian-novikov-db any idea why this snippet came through ? Is that actually a needed change in `maven-core` ?

Reply:
> @maximilian-novikov-db any idea why this snippet came through ? Is that actually a needed change in `maven-core` ?

Because there is no any sense to call this method when pluginRealm is not set in plugin descriptor 
this code is guaranteed to fail with a NPE in that case
I've raised PR https://github.com/apache/maven/pull/612 to this branch please take a look how this method is used in CachingMojosExecutionStrategy
This allows us to get rid of MojoChecker and related functionality